modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/RegexReplace.java

```diff
@@ -0,0 +1,173 @@
+package org.daisy.pipeline.tts.cereproc.impl.util;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.Vector;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+/**
+ * A class providing simple regex search and replace.
+ * A rule-file is parsed and regexes and their replacement
+ * string are stored in an ordered collection.
+ * This Class has been migrated from Pipeline 1.
+ * @author Martin Blomberg
+ * @author carl walinder
+ *
+ */
+public class RegexReplace {
+
+    private XPath xpathObj = XPathFactory.newInstance().newXPath();
+    private Vector<PatternReplace> patterns = new Vector<PatternReplace>();
+    private boolean DEBUG = false;
+
+    /**
+     * @param rulesXML
+     */
+    public RegexReplace(URL rulesXML) {
+        initResources(rulesXML);
+    }
+
+
+    /**
+     * Reads the config file containing regexes and replace strings.
+     * @param rulesXML the config file.
+     */
+    private void initResources(URL rulesXML) {
+        try {
+            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+            DocumentBuilder db = dbf.newDocumentBuilder();
+            String xpath = "//rule";
+
+            Document config = db.parse(rulesXML.openStream());
+            NodeList rules = this.selectNodes(config.getDocumentElement(), xpath);
+
+            for (int i = 0; i < rules.getLength(); i++) {
+                Element rule = (Element) rules.item(i);
+                try {
+                    Pattern p = Pattern.compile(rule.getAttribute("match"));
+                    PatternReplace pr = new PatternReplace(p, rule.getAttribute("replace"));
+                    patterns.add(pr);
+                    DEBUG(pr);
+                } catch(PatternSyntaxException pse) {
+                    System.out.println("There is a problem with the regular expression!");
+                    System.out.println("The pattern in question is: "+pse.getPattern());
+                    System.out.println("The description is: "+pse.getDescription());
+                    System.out.println("The message is: "+pse.getMessage());
+                    System.out.println("The index is: "+pse.getIndex());
```

Comment:
Maybe these can become log messages? Or the text can be included in the IllegalArgumentException? Same for the other `System.err.println` below.

Reply:
I changed it to log messages.

Reply:
Still one System.err.println() left. Did you miss it or leave it on purpose?

Reply:
hmm.. I thought i went threw them all. I must have missed that one.

Reply:
Changed them to log messages now.

Reply:
> OK looks good. Now there is another thing left. Currently the XSLT will be executed twice. One by the px:ssml-to-audio step and once again by CereProcEngine. I will change px:ssml-to-audio so that it accepts TTSServices that return null for getSSMLxslTransformerURL().

Ok! Will you make the change to this PR och should this me merged first?

Reply:
I can't push to your branch but I could add it to my [other PR](https://github.com/mtmse/pipeline/pull/13)? Or I can push it to another branch that you can cherry-pick from. Let me know what you prefer.

Reply:
Ok, i can cherry-pick it if you give me the commit hash :)

Reply:
https://github.com/daisy/pipeline/compare/optional-ssml-transformer^^...optional-ssml-transformer. There are two commits: the one that makes the SSML transformer optional, and another commit that I had to include to avoid conflicts. This commit will be useful as well. I mentioned it on one of the previous calls.